### PR TITLE
feat: reorder and delete functionality for question editor

### PR DESCRIPTION
**Pull request by joyce-shi**
## Notion ticket link

[Add context for Question Editor](https://www.notion.so/uwblueprintexecs/Add-context-for-Question-Editor-bfe32be430fa45bebc08a39e2e041d6a?pvs=4)
[Delete functionality in Question Creation module](https://www.notion.so/uwblueprintexecs/Delete-functionality-in-Question-Creation-module-799e35eefc33415aacba1e2003928f35?pvs=4)
[Reorder functionality in Question Creation module](https://www.notion.so/uwblueprintexecs/Add-context-for-Question-Editor-bfe32be430fa45bebc08a39e2e041d6a?pvs=4)

## Implementation description
* refactored existing question editor to use context to manage state
* added functionality to delete question elements
* added functionality to reorder question elements
Note: I didn't implement some of the design hover behaviour, since I wanted to get the basic functionality merged first. I'll come back to that in a bit!

## What should reviewers focus on?
* reorder works correctly
* delete works correctly
* no infinite loops

### CODE REVIEW

File: frontend/src/components/question-creation/QuestionElementItem.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      const hoverClientY = (clientOffset as XYCoord).y - hoverBoundingRect.top;
+      if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
+        return;
+      }
+      if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
+        return;
+      }
+      reorderQuestionElements(dragIndex, hoverIndex);
+      /* eslint-disable no-param-reassign */
+      item.index = hoverIndex;
```
Comment (joyce-shi) [inline frontend/src/components/question-creation/QuestionElementItem.tsx]: documentation recommends ignoring `no-param-reassign` in this case for performance reasons

File: frontend/src/components/question-creation/QuestionElementItem.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import { DeleteOutlineIcon, HamburgerMenuIcon } from "../../assets/icons";
+import {
+  QuestionElement,
+  QuestionElementType,
+} from "../../types/QuestionTypes";
+import TextElement from "./question-elements/TextElement";
+import QuestionEditorContext from "../../contexts/QuestionEditorContext";
+import { DragTypes } from "../../types/DragTypes";
+
+export interface QuestionElementItemProps {
```
Comment (carissa-tang) [inline frontend/src/components/question-creation/QuestionElementItem.tsx]:
```suggestion
interface QuestionElementItemProps {
```

File: frontend/src/components/question-creation/QuestionElementItem.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    { handlerId: Identifier | null }
+  >({
+    accept: DragTypes.QUESTION_ELEMENT_ITEM,
+    collect(monitor) {
+      return {
+        handlerId: monitor.getHandlerId(),
+      };
+    },
+    hover(item: DragItem, monitor) {
+      if (!previewRef.current) {
```
Comment (carissa-tang) [inline frontend/src/components/question-creation/QuestionElementItem.tsx:79]: hmm can we maybe refactor some of this logic into a function and put in `/utils`? just to clean up a bit, but if it makes code more complex then it's fine as it

**Review by carissa-tang — APPROVED** [review]
lgtm!!!

File: frontend/src/components/question-creation/QuestionElementItem.tsx (hunk 3 above)
Comment (joyce-shi) [reply frontend/src/components/question-creation/QuestionElementItem.tsx:79]: good idea! created `Question Utils` 🤠

**Pull request merged**